In [2]:
import polars as pl
import pandas as pd
from IPython.display import display



In [3]:
from PbDriver.features import validate_features_vs_target

## Load 

Load from CSV and clean up the data
1. Remove all rows where the year is 2014 because its the first year there are no pbs
2. Clean up gender as it contains "M" or "Male" and "F" or "Female". Just take the first character for simplity
3. Data Types

In [4]:
import PbDriver.DataPrep as dp
results_with_pb_df = dp.get_data()

display(results_with_pb_df)




shape: (55_523, 65)
┌──────────┬──────────┬───────────────┬────────────┬───┬─────────────┬─────────┬─────────┬─────────┐
│ Position ┆ Time     ┆ time_in_secon ┆ Bib Number ┆ … ┆ average_age ┆ min_dob ┆ max_dob ┆ avg_dob │
│ ---      ┆ ---      ┆ ds            ┆ ---        ┆   ┆ ---         ┆ ---     ┆ ---     ┆ ---     │
│ i64      ┆ str      ┆ ---           ┆ i64        ┆   ┆ f64         ┆ f64     ┆ f64     ┆ f64     │
│          ┆          ┆ f64           ┆            ┆   ┆             ┆         ┆         ┆         │
╞══════════╪══════════╪═══════════════╪════════════╪═══╪═════════════╪═════════╪═════════╪═════════╡
│ 3        ┆ 00:29:14 ┆ 1754.0        ┆ 880        ┆ … ┆ 42.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 4        ┆ 00:29:34 ┆ 1774.0        ┆ 588        ┆ … ┆ 52.0        ┆ 1964.0  ┆ 1960.0  ┆ 1962.0  │
│ 5        ┆ 00:29:47 ┆ 1787.0        ┆ 902        ┆ … ┆ 37.0        ┆ 1979.0  ┆ 1975.0  ┆ 1977.0  │
│ 6        ┆ 00:30:07 ┆ 1807.0        ┆ 918        ┆ … ┆ 18.5        ┆ 2001.0  ┆ 1990.0  ┆ 1995.5  │
│ 7        ┆ 00:30:09 ┆ 1809.0        ┆ 530        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ …        ┆ …        ┆ …             ┆ …          ┆ … ┆ …           ┆ …       ┆ …       ┆ …       │
│ 627      ┆ 00:47:41 ┆ 2861.0        ┆ 631        ┆ … ┆ 59.0        ┆ 1959.0  ┆ 1955.0  ┆ 1957.0  │
│ 628      ┆ 00:54:49 ┆ 3289.0        ┆ 613        ┆ … ┆ 64.0        ┆ 1954.0  ┆ 1950.0  ┆ 1952.0  │
│ 629      ┆ 00:56:27 ┆ 3387.0        ┆ 429        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 630      ┆ 00:57:56 ┆ 3476.0        ┆ 151        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 631      ┆ 00:58:30 ┆ 3510.0        ┆ 606        ┆ … ┆ 49.0        ┆ 1969.0  ┆ 1965.0  ┆ 1967.0  │
└──────────┴──────────┴───────────────┴────────────┴───┴─────────────┴─────────┴─────────┴─────────┘

## Do we have balanced target classes?

We calculate the PB from the prior runs. 
- The first year is 2014 which has no pbs because there is no prior data
- This needs to be removed as its unbalances the data

We also need to remove any first time only runs as a first is always a pb (technically) and will unbalance the data

In [5]:
results_with_pb_df_all = results_with_pb_df

def pivot_on_year(df: pl.DataFrame):
    df_first_timers = df.filter(pl.col("first_time_run") == True)

    number_of_first_time_runs = len(df_first_timers)
    if number_of_first_time_runs > 0:
        print(f"WARNING this data set includes {number_of_first_time_runs} first time runs that will never be a PB")
    else:
        print("PERFECT No first time runs in the data set")
    
    print(f"Total number of rows: {len(df)}")
    
    # Pivot the DataFrame
    pivot_df = df.pivot(
        values=["is_pb"],
        index="year",
        on="is_pb",
        aggregate_function="len"
    ).sort("year")
    
    display(pivot_df)



# Remove all rows where the year is 2014
results_with_pb_df_filtered = results_with_pb_df.filter(
    (pl.col("year") != 2014) & (pl.col("first_time_run") == False)
)

print ('Before')

pivot_on_year(results_with_pb_df_all)

print ('After - without 2014 and first time runs == false')

pivot_on_year(results_with_pb_df_filtered)


Before
WARNING this data set includes 33080 first time runs that will never be a PB
Total number of rows: 55523


shape: (10, 3)
┌──────┬───────┬──────┐
│ year ┆ false ┆ true │
│ ---  ┆ ---   ┆ ---  │
│ i64  ┆ u32   ┆ u32  │
╞══════╪═══════╪══════╡
│ 2014 ┆ 6811  ┆ null │
│ 2015 ┆ 6289  ┆ 1116 │
│ 2016 ┆ 5925  ┆ 1283 │
│ 2017 ┆ 5263  ┆ 1042 │
│ 2018 ┆ 5269  ┆ 904  │
│ 2019 ┆ 5451  ┆ 1172 │
│ 2020 ┆ 378   ┆ 93   │
│ 2022 ┆ 3967  ┆ 637  │
│ 2023 ┆ 3758  ┆ 631  │
│ 2024 ┆ 4938  ┆ 596  │
└──────┴───────┴──────┘

After - without 2014 and first time runs == false
PERFECT No first time runs in the data set
Total number of rows: 22443


shape: (9, 3)
┌──────┬──────┬───────┐
│ year ┆ true ┆ false │
│ ---  ┆ ---  ┆ ---   │
│ i64  ┆ u32  ┆ u32   │
╞══════╪══════╪═══════╡
│ 2015 ┆ 1116 ┆ 1251  │
│ 2016 ┆ 1283 ┆ 1821  │
│ 2017 ┆ 1042 ┆ 2023  │
│ 2018 ┆ 904  ┆ 2036  │
│ 2019 ┆ 1172 ┆ 2428  │
│ 2020 ┆ 93   ┆ 167   │
│ 2022 ┆ 637  ┆ 1536  │
│ 2023 ┆ 631  ┆ 1605  │
│ 2024 ┆ 596  ┆ 2102  │
└──────┴──────┴───────┘

## Find Categoricals

These are columns with limited subset of values.

For example **education** could be 
- Graduate
- Not Graduate

In our data set
- Club 
- RaceName
- Gender
All of which are unlikely to influence a pb

However, I'll add of these are categorical variables to the model.
- RaceName
- Gender

The two `to_dummies` colum will turn these into columns with 1s and 0s.
- So for `RaceName` it will create a column for each unique value in the column and set it to 1 if the value is present and 0 if not.
- `RaceName_Downland Dash`
- `RaceName_Fittleworth`
- for each race combo





In [6]:
results_with_pb_with_dummies_df = results_with_pb_df_filtered.to_dummies(
    ["Gender", "RaceName", 'p_of_10_road_age_category','is_club_member'], drop_first=True
)


In [5]:
display(results_with_pb_with_dummies_df)

shape: (22_443, 102)
┌──────────┬──────────┬───────────────┬────────────┬───┬─────────────┬─────────┬─────────┬─────────┐
│ Position ┆ Time     ┆ time_in_secon ┆ Bib Number ┆ … ┆ average_age ┆ min_dob ┆ max_dob ┆ avg_dob │
│ ---      ┆ ---      ┆ ds            ┆ ---        ┆   ┆ ---         ┆ ---     ┆ ---     ┆ ---     │
│ i64      ┆ str      ┆ ---           ┆ i64        ┆   ┆ f64         ┆ f64     ┆ f64     ┆ f64     │
│          ┆          ┆ f64           ┆            ┆   ┆             ┆         ┆         ┆         │
╞══════════╪══════════╪═══════════════╪════════════╪═══╪═════════════╪═════════╪═════════╪═════════╡
│ 7        ┆ 00:29:11 ┆ 1751.0        ┆ 167        ┆ … ┆ 53.0        ┆ 1964.0  ┆ 1960.0  ┆ 1962.0  │
│ 13       ┆ 00:29:54 ┆ 1794.0        ┆ 203        ┆ … ┆ 28.0        ┆ 1989.0  ┆ 1985.0  ┆ 1987.0  │
│ 19       ┆ 00:30:24 ┆ 1824.0        ┆ 294        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 21       ┆ 00:30:29 ┆ 1829.0        ┆ 390        ┆ … ┆ 43.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 24       ┆ 00:30:45 ┆ 1845.0        ┆ 295        ┆ … ┆ 28.0        ┆ 1989.0  ┆ 1985.0  ┆ 1987.0  │
│ …        ┆ …        ┆ …             ┆ …          ┆ … ┆ …           ┆ …       ┆ …       ┆ …       │
│ 615      ┆ 00:45:24 ┆ 2724.0        ┆ 203        ┆ … ┆ 44.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 616      ┆ 00:45:37 ┆ 2737.0        ┆ 112        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 617      ┆ 00:46:14 ┆ 2774.0        ┆ 63         ┆ … ┆ 34.0        ┆ 1984.0  ┆ 1980.0  ┆ 1982.0  │
│ 623      ┆ 00:49:33 ┆ 2973.0        ┆ 42         ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 627      ┆ 00:47:41 ┆ 2861.0        ┆ 631        ┆ … ┆ 59.0        ┆ 1959.0  ┆ 1955.0  ┆ 1957.0  │
└──────────┴──────────┴───────────────┴────────────┴───┴─────────────┴─────────┴─────────┴─────────┘

## Split the data

Remove the target column from the features. **is_pb**

Features are the columns that will be used to predict the target variable. 

In this case, the features are all the columns that **Known before the race starts**
- Prior performance 
- Gender
- Club mates (you would see them before the start and on course that could inspire you to run faster)
 


In [7]:

from PbDriver.features import filter_features

pb_achieved_features_df = filter_features(results_with_pb_with_dummies_df)

pb_achieved_target_df = results_with_pb_df_filtered.select("is_pb")

validate_features_vs_target(pb_achieved_features_df,pb_achieved_target_df)


display(pb_achieved_features_df)



MATCH Features_df shape: (22443, 62) vs Target (22443, 1)


shape: (22_443, 62)
┌──────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ Gender_F ┆ is_club_m ┆ RaceName_ ┆ RaceName_ ┆ … ┆ p_of_10_r ┆ p_of_10_r ┆ p_of_10_r ┆ average_a │
│ ---      ┆ ember_fal ┆ Downland  ┆ Fittlewor ┆   ┆ oad_age_c ┆ oad_age_c ┆ oad_age_c ┆ ge        │
│ u8       ┆ se        ┆ Dash      ┆ th        ┆   ┆ ategory_V ┆ ategory_V ┆ ategory_V ┆ ---       │
│          ┆ ---       ┆ ---       ┆ ---       ┆   ┆ 80        ┆ 85        ┆ 95        ┆ f64       │
│          ┆ u8        ┆ u8        ┆ u8        ┆   ┆ ---       ┆ ---       ┆ ---       ┆           │
│          ┆           ┆           ┆           ┆   ┆ u8        ┆ u8        ┆ u8        ┆           │
╞══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 53.0      │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 28.0      │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ null      │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 43.0      │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 28.0      │
│ …        ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 44.0      │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ null      │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 34.0      │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ null      │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 59.0      │
└──────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

SEED = 123

(
    pb_achieved_feature_train_df,
    pb_achieved_feature_test_df,
    pb_achieved_target_train_df,
    pb_achieved_target_test_df,
) = train_test_split(
    pb_achieved_features_df,
    pb_achieved_target_df,
    test_size=0.30,
    random_state=SEED,
)

total_rows = len(pb_achieved_features_df)
train_rows = len(pb_achieved_feature_train_df)
test_rows = len(pb_achieved_feature_test_df)

test_percentage_of_total = train_rows / total_rows

print(f"Total Rows: {total_rows} Training Rows: {train_rows} Test Rows: {test_rows} Test Percentage of Total: {test_percentage_of_total:.2f}")

validate_features_vs_target(pb_achieved_feature_train_df,pb_achieved_target_train_df)
validate_features_vs_target(pb_achieved_feature_test_df,pb_achieved_target_test_df)


Total Rows: 22443 Training Rows: 15710 Test Rows: 6733 Test Percentage of Total: 0.70
MATCH Features_df shape: (15710, 62) vs Target (15710, 1)
MATCH Features_df shape: (6733, 62) vs Target (6733, 1)


## Scale the data

It is important to scale the data so that the model is not distorted by really large values.

### My notes are 
- Dimensionality reduction, which 
1. Rotates the data. Some like 45 degrees
2. Keeps the same data but reduce the complexity
- Principal Component Analysis (PCA) is a technique that can be used to reduce the dimensionality of the data.

### Notes from Copilot
The StandardScaler standardizes a feature by subtracting the mean and then scaling to unit variance. Unit variance means dividing all the values by the standard deviation.

In [ ]:
scaler = StandardScaler()

pb_achieved_feature_train_scaled_df = scaler.fit_transform(
    pb_achieved_feature_train_df
)

pb_achieved_feature_test_scaled_df = scaler.fit_transform(
    pb_achieved_feature_test_df
)

print(f"Scaled Training data shape: {pb_achieved_feature_train_scaled_df.shape} vs Test {pb_achieved_feature_test_scaled_df.shape}")



Do we have balanced target classes?

# Logistic regression example

```
from sklearn.linear_model import LogisticRegression

logistic_regression_classifier = LogisticRegression(penalty=None)

logistic_regression_classifier.fit(
    pb_achieved_feature_train_scaled_df,
    pb_achieved_target_train_df.get_column("is_pb"),
)
```

## Constraints

> ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values


One common strategy is to use the mean or median of the column to fill in the missing values. Here is how you can do it using SimpleImputer from sklearn.impute:  
Import SimpleImputer.
Create an imputer instance with the desired strategy (e.g., mean, median).
Fit the imputer on the training data and transform both the training and test data.

```
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression

# Create an imputer instance with the mean strategy
imputer = SimpleImputer(strategy="mean")

# Fit the imputer on the training data and transform both the training and test data
pb_achieved_feature_train_scaled_df = imputer.fit_transform(pb_achieved_feature_train_scaled_df)
pb_achieved_feature_test_scaled_df = imputer.transform(pb_achieved_feature_test_scaled_df)

# Create and fit the logistic regression classifier
logistic_regression_classifier = LogisticRegression(penalty=None)

logistic_regression_classifier.fit(
    pb_achieved_feature_train_scaled_df,
    pb_achieved_target_train_df.get_column("is_pb"),
)
```





In [10]:
# Convert Polars DataFrame to Pandas DataFrame
pb_achieved_feature_train_df_pandas = pb_achieved_feature_train_df.to_pandas()
pb_achieved_target_train_df_pandas = pb_achieved_target_train_df.to_pandas()
pb_achieved_feature_test_df_pandas = pb_achieved_feature_test_df.to_pandas()
pb_achieved_target_test_df_pandas = pb_achieved_target_test_df.to_pandas()



# Define Column for Defaults

**All columns** need **default values** for **missing values**.

This can have quite an impact on the model. For example if 
- You defaulted missing **points to 10** which is the **highest point** you can score 
- You would **create an artificially high amount** of high points scores when the data is much lower

The second section of the code checks the column exist in the data frame. If they do not exist you will get an error.

> ValueError: A given column is not a column of the dataframe 

In [11]:

def get_boolean_features(df : pd.DataFrame, exclude_races = True, exclude_ages = True) -> list:
    binary_columns = []
    
    for col in df.columns:
        if exclude_races is False and col.startswith('RaceName_'):
            binary_columns.append(col)
        elif exclude_ages is False and col.startswith('p_of_10_road_age_category_'):
            binary_columns.append(col)
     
    binary_columns.append("is_club_member_false")
    binary_columns.append("Gender_F")
    
    return binary_columns


In [39]:
binary_columns = get_boolean_features(pb_achieved_feature_train_df_pandas, True, True)

print(binary_columns)

['is_club_member_false', 'Gender_F']


In [40]:
def validate_features(features_df, 
                      low_value_good_features: list,
                      high_value_good_features: list,
                      boolean_features: list,
                      time_in_seconds_features: list,
                      years_since_value_features: list,
                      average_ok_columns) -> list:
    
  
    feature_columns_to_validate = (low_value_good_features 
                                      + high_value_good_features 
                                      + boolean_features 
                                      + time_in_seconds_features 
                                      + years_since_value_features
                                      + average_ok_columns)
  

    # Find columns that are not in the combined feature list
    extra_columns = [col for col in features_df.columns if col not in feature_columns_to_validate]
    print(f"WARNING (Be Aware)- Column that exist in the features_df but not in the feature_columns_to_validate list: {extra_columns}")
    
    # Check if columns exist in the DataFrame
    missing_columns = [col for col in feature_columns_to_validate if col not in features_df.columns]
    missing_low_value_good_columns = [col for col in low_value_good_features if col not in features_df.columns]
    missing_high_value_good_columns = [col for col in high_value_good_features if col not in features_df.columns]
    
    good_prefix ="GOOD"
    bad_prefix ="BAD"
    
    
    # Print the results
    if missing_columns:
        print(f"{bad_prefix} - Missing columns in WHOLE set: {missing_columns}")
    else:
        print(f"{good_prefix} - All feature columns {len(feature_columns_to_validate)} exist in the DataFrame {len(features_df.columns)}.")
    
    if missing_low_value_good_columns:
        print(f"{bad_prefix} - Missing columns in low_value_good_columns: {missing_low_value_good_columns}")
    else:
        print(f"{good_prefix} - All columns in low_value_good_columns exist in the DataFrame.")
    
    if missing_high_value_good_columns:
        print(f"{bad_prefix} - Missing columns in high_value_good_columns: {missing_high_value_good_columns}")
    else:
        print(f"{good_prefix} - All columns in high_value_good_columns exist in the DataFrame.")
    
    
    # Create a DataFrame with the count of NaN values for each column
    nan_counts_df = pd.DataFrame({
        'column': features_df.columns,
        'nan_count': features_df.isna().sum()
    })
    
    # Sort the DataFrame by nan_count in descending order
    nan_counts_df = nan_counts_df.sort_values(by='nan_count', ascending=False)
    nan_counts_df = nan_counts_df[nan_counts_df['nan_count'] > 0]
    
    # Set the maximum number of rows to display
    pd.set_option('display.max_rows', 30)
    
    display(nan_counts_df.head(8))
    
    return feature_columns_to_validate

In [41]:

# Define the columns
# "Position", "Gender Position", 
time_in_seconds_columns = [ "prior_quickest_run_time_in_seconds"]

years_since_value_columns = ["position_record_years_since_best", "position_record_years_since_worst"]

average_ok_columns = ["average_age"]

low_value_good_columns = [
    "best_position_year_to_date", "average_position_year_to_date", "worst_position_year_to_date", "position_record_best_position_all_time", "position_record_worst_position_all_time", "prior_quickest_run_position",  "position_record_average_position_all_time"] 


high_value_good_columns = ["fellow_club_mates_running", "fellow_club_mates_running_male", "fellow_club_mates_running_female", "average_points_year_to_date", "best_points_year_to_date", "worst_points_year_to_date", "runs_in_wsfrl_year_to_date", 'wsfrl_tenure_in_runs','wsfrl_tenure_in_years', 'prior_run_attempts']








In [42]:
# Combine the lists
all_feature_columns = validate_features(pb_achieved_feature_train_df_pandas,
                                        low_value_good_columns, 
                                        high_value_good_columns, 
                                        binary_columns, 
                                        time_in_seconds_columns, 
                                        years_since_value_columns,
                                        average_ok_columns)


WARNING (Be Aware)- Column that exist in the features_df but not in the feature_columns_to_validate list: ['RaceName_Downland Dash', 'RaceName_Fittleworth', 'RaceName_Great Walstead 5', 'RaceName_Gunpowder Trot', 'RaceName_Hangover-5', 'RaceName_Hedgehoppers 5', 'RaceName_Hickstead Gallop', 'RaceName_Highdown Hike', 'RaceName_Hornets Stinger', 'RaceName_Hove Park', 'RaceName_Lewes AC Easter Monday', 'RaceName_Run Your Heart Out 8k', 'RaceName_Seven Stiles', 'RaceName_Steepdown Challenge', 'RaceName_Steyning AC Roundhill Romp', 'RaceName_Tilgate 5', 'RaceName_Trundle View', 'RaceName_Valentine', 'RaceName_Windlesham-Whip', 'RaceName_Worthing Lido', 'p_of_10_road_age_category_SEN', 'p_of_10_road_age_category_TBC', 'p_of_10_road_age_category_U11', 'p_of_10_road_age_category_U13', 'p_of_10_road_age_category_U15', 'p_of_10_road_age_category_U17', 'p_of_10_road_age_category_U20', 'p_of_10_road_age_category_U23', 'p_of_10_road_age_category_V35', 'p_of_10_road_age_category_V40', 'p_of_10_road_

column  nan_count
average_age                                      average_age       2944
best_points_year_to_date            best_points_year_to_date       1626
average_points_year_to_date      average_points_year_to_date       1626
worst_points_year_to_date          worst_points_year_to_date       1626
best_position_year_to_date        best_position_year_to_date       1080
average_position_year_to_date  average_position_year_to_date       1080
worst_position_year_to_date      worst_position_year_to_date       1080

In [43]:
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Create imputers for different columns

low_value_imputer = SimpleImputer(strategy="constant", fill_value=400)
high_value_imputer = SimpleImputer(strategy="constant", fill_value=0) #-np.inf is too big for a float
binary_value_imputer = SimpleImputer(strategy="constant", fill_value=np.uint8(0))
seconds_value_imputer = SimpleImputer(strategy="constant", fill_value=400000)
average_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')


# Combine the imputers using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ("low_value_good", low_value_imputer, low_value_good_columns),
        ("high_value_good", high_value_imputer, high_value_good_columns),
        ("binary", binary_value_imputer, binary_columns),
        ("seconds", seconds_value_imputer, time_in_seconds_columns),
        ("average", average_imputer, average_ok_columns),
        ("years_since", SimpleImputer(strategy="constant", fill_value=0), years_since_value_columns),
    ]
)

PIPELINE_STEP_PREPROCESSOR = "preprocessor"
PIPELINE_STEP_SCALER = "scaler"

# Create a pipeline with the preprocessor, scaler, and the logistic regression classifier
pipeline = Pipeline(steps=[
    (PIPELINE_STEP_PREPROCESSOR, preprocessor),
    (PIPELINE_STEP_SCALER, StandardScaler()),
    ("classifier", LogisticRegression(penalty=None))
])

# Fit the pipeline on the training data
pipeline.fit(pb_achieved_feature_train_df_pandas, pb_achieved_target_train_df_pandas["is_pb"])

# 
# # Transform the test data
# pb_achieved_feature_test_scaled_df = pipeline.transform(pb_achieved_feature_test_df)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('low_value_good',
                                                  SimpleImputer(fill_value=400,
                                                                strategy='constant'),
                                                  ['best_position_year_to_date',
                                                   'average_position_year_to_date',
                                                   'worst_position_year_to_date',
                                                   'position_record_best_position_all_time',
                                                   'position_record_worst_position_all_time',
                                                   'prior_quickest_run_position',
                                                   'position_rec...
                                                  SimpleImputer(fill_value=400000,
                                                                strategy='constant'),
                                                  ['prior_quickest_run_time_in_seconds']),
                                                 ('average', SimpleImputer(),
                                                  ['average_age']),
                                                 ('years_since',
                                                  SimpleImputer(fill_value=0,
                                                                strategy='constant'),
                                                  ['position_record_years_since_best',
                                                   'position_record_years_since_worst'])])),
                ('scaler', StandardScaler()),
                ('classifier', LogisticRegression(penalty=None))])

## Run Pipeline Stages explicitly

To run the full pipeline 
`pb_achieved_feature_test_scaled_df = pipeline.transform(pb_achieved_feature_test_df)`

However this does not work 
> AttributeError: This 'Pipeline' has no attribute 'transform'

this is because `LogisticRegression` does not have a `transform` method, but a `fit`. 
  


In [45]:
from sklearn.linear_model import LogisticRegression


def run_pipeline_in_stages(df_feature : pd.DataFrame, df_target : pd.DataFrame):

    # Transform the test data using the preprocessor part of the pipeline
    preprocessed_df = pipeline.named_steps['preprocessor'].transform(df_feature)

    # Scale the preprocessed test data
    scaled_df = pipeline.named_steps['scaler'].transform(preprocessed_df)
    
    classifier = LogisticRegression(penalty=None)

    classifier.fit(
        scaled_df,
        df_target["is_pb"],
    )
    
    return scaled_df, classifier 
    

In [46]:
(scaled_train_df,
logistic_regression_classifier) = run_pipeline_in_stages(
    pb_achieved_feature_train_df_pandas,
    pb_achieved_target_train_df_pandas)





In [47]:
(scaled_test_df,
logistic_regression_classifier_junk) = run_pipeline_in_stages(
    pb_achieved_feature_test_df_pandas,
    pb_achieved_target_test_df_pandas)

In [28]:


logistic_regression_classifier.classes_



array([False,  True])

In [48]:
import altair as alt

def display_feature_coefficients(classifier, feature_columns):
    # Convert the set to a list to maintain order
    all_feature_columns_list = list(feature_columns)
    
    
    # Create a DataFrame with the features and their coefficients
    feature_coefficients_df = pl.DataFrame(
        {
            "feature": all_feature_columns_list,
            "coefficient": classifier.coef_[0][:len(feature_columns)],
        }
    )
    
    # Sort the DataFrame by coefficient in descending order
    sorted_feature_coefficients_df = feature_coefficients_df.sort("coefficient", descending=True)
    
    # Plot the bar chart
    alt.Chart(sorted_feature_coefficients_df.to_pandas()).mark_bar().encode(
        x="coefficient",
        y=alt.Y("feature", sort="-x")
    ).properties(
        width=500,
        height=300,
    ).display()

In [49]:
display_feature_coefficients(logistic_regression_classifier, all_feature_columns)

alt.Chart(...)

In [50]:
logistic_regression_classifier.coef_

array([[ 0.14072087, -0.16021094, -0.22367167,  0.07342591,  0.1219994 ,
         0.36082159, -0.37284477, -0.020228  , -0.0332092 , -0.00344088,
         0.31791096,  0.04074353, -0.16924722,  0.16054202, -0.27150845,
        -0.16248632, -0.48311617,  0.02797542, -0.06718529,  3.20872515,
        -0.17295688, -0.38744207,  0.20488273]])

In [53]:

# Make predictions on the test data
predictions = pipeline.named_steps['classifier'].predict(scaled_train_df)

# Print the predictions
print(predictions)

[False False  True ... False False False]


In [54]:
validate_features_vs_target(pb_achieved_feature_train_df_pandas, pb_achieved_target_train_df_pandas)
validate_features_vs_target(scaled_train_df, pb_achieved_target_train_df_pandas)

validate_features_vs_target(pb_achieved_feature_test_df_pandas, pb_achieved_target_test_df_pandas)

MATCH Features_df shape: (15710, 62) vs Target (15710, 1)
MATCH Features_df shape: (15710, 23) vs Target (15710, 1)
MATCH Features_df shape: (6733, 62) vs Target (6733, 1)


In [ ]:
> What are the probabilities that each of the observations in the test dataset have been assigned to the classes?

In [ ]:

logistic_regression_classifier.predict_proba(scaled_test_df)

In [ ]:
logistic_regression_classifier.predict(scaled_test_df)



```
ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values
```

In [55]:
from sklearn.metrics import classification_report

validate_features_vs_target(scaled_test_df, pb_achieved_target_test_df)


print(
    classification_report(
        pb_achieved_target_test_df.get_column("is_pb"),
        logistic_regression_classifier.predict(scaled_test_df),
    )
)

MATCH Features_df shape: (6733, 23) vs Target (6733, 1)
              precision    recall  f1-score   support

       False       0.73      0.87      0.80      4468
        True       0.59      0.37      0.46      2265

    accuracy                           0.70      6733
   macro avg       0.66      0.62      0.63      6733
weighted avg       0.69      0.70      0.68      6733



## Interpreting Results from the Classification Report

The classification report provides a summary of performance metrics for a classification model. Below is the detailed breakdown and interpretation:

### Definitions
1. **Precision**: The ratio of true positives to total predicted positives. Indicates how many of the predicted "True" labels are actually "True."
   \[
   \text{Precision} = \frac{\text{True Positives}}{\text{True Positives + False Positives}}
   \]

2. **Recall (Sensitivity)**: The ratio of true positives to total actual positives. Measures the ability of the classifier to identify all "True" instances.
   \[
   \text{Recall} = \frac{\text{True Positives}}{\text{True Positives + False Negatives}}
   \]

3. **F1-Score**: The harmonic mean of precision and recall, balancing their trade-off.
   \[
   \text{F1-Score} = 2 \cdot \frac{\text{Precision} \cdot \text{Recall}}{\text{Precision + Recall}}
   \]

4. **Support**: The number of actual instances in each class.

---

### Interpreting the Results

#### **Class: False (Negative Class)**
- **Precision**: `0.74`  
   - 74% of the predictions for "False" were correct.
- **Recall**: `0.88`  
   - 88% of the actual "False" cases were correctly identified by the model.
- **F1-Score**: `0.80`  
   - A strong balance between precision and recall for this class.
- **Support**: `4468`  
   - There are 4468 instances of "False" in the dataset.

#### **Class: True (Positive Class)**
- **Precision**: `0.62`  
   - 62% of the predictions for "True" were correct.
- **Recall**: `0.38`  
   - Only 38% of the actual "True" cases were correctly identified.
- **F1-Score**: `0.47`  
   - A relatively low score, indicating that the model struggles with this class.
- **Support**: `2265`  
   - There are 2265 instances of "True" in the dataset.

---

### Overall Metrics
- **Accuracy**: `0.71`  
   - The model correctly classified 71% of all instances.
- **Macro Average**:
   - **Precision**: `0.68`  
   - **Recall**: `0.63`  
   - **F1-Score**: `0.64`  
   - These averages give equal weight to both classes, regardless of support.
- **Weighted Average**:
   - **Precision**: `0.70`  
   - **Recall**: `0.71`  
   - **F1-Score**: `0.69`  
   - These averages take into account the support of each class, so the "False" class has more influence due to its larger support.

---

### Insights
1. **Class Imbalance**:  
   - There is a class imbalance, with more "False" instances (`4468`) than "True" (`2265`). This likely affects the model's ability to detect the minority class ("True").

2. **Performance on "True" Class**:
   - The model struggles to correctly identify the "True" class (low recall of `0.38`). This might indicate a need for techniques to handle class imbalance, such as oversampling, undersampling, or class weighting.

3. **Overall Performance**:
   - While the model performs reasonably well on the majority class ("False"), its performance on the minority class ("True") drags down the overall metrics.

---

### Recommendations
1. **Address Class Imbalance**:  
   - Consider using techniques like SMOTE, ADASYN, or class weighting in the logistic regression model to improve recall for the "True" class.

2. **Threshold Tuning**:  
   - Experiment with adjusting the decision threshold of the model to better balance precision and recall.

3. **Model Complexity**:  
   - Consider using a more complex model (e.g., ensemble methods like Random Forest or Gradient Boosting) if the dataset supports it.

4. **Feature Engineering**:  
   - Investigate additional features or transformations to enhance separability between the classes.

---
